controllers/k8ssandra/cassandra_telemetry_reconciler.go

```diff
@@ -0,0 +1,94 @@
+// Logic in this file reconciles telemetry resources into the state declared in the CRs.
+
+package k8ssandra
+
+import (
+	"context"
+	"errors"
+	"github.com/go-logr/logr"
+	cassdcapi "github.com/k8ssandra/cass-operator/apis/cassandra/v1beta1"
+	api "github.com/k8ssandra/k8ssandra-operator/apis/k8ssandra/v1alpha1"
+	k8ssandraapi "github.com/k8ssandra/k8ssandra-operator/apis/k8ssandra/v1alpha1"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/result"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/telemetry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func (r *K8ssandraClusterReconciler) reconcileCassandraDCTelemetry(
+	ctx context.Context,
+	kc *api.K8ssandraCluster,
+	dcTemplate api.CassandraDatacenterTemplate,
+	actualDc *cassdcapi.CassandraDatacenter,
+	logger logr.Logger,
+	remoteClient client.Client,
+) result.ReconcileResult {
+	logger.Info("reconciling telemetry")
+	mergedSpec := kc.Spec.Cassandra.CassandraTelemetry.Merge(dcTemplate.CassandraTelemetry)
+	cfg := telemetry.PrometheusResourcer{
+		MonitoringTargetNS:   actualDc.Namespace,
+		MonitoringTargetName: actualDc.Name,
+		ServiceMonitorName:   kc.Name + "-" + actualDc.Name + "-" + "cass-servicemonitor",
+		Logger:               logger,
+		CommonLabels:         mustLabels(kc.Name, actualDc.Name),
+	}
+	logger.Info("merged TelemetrySpec constructed", "mergedSpec", mergedSpec, "cluster", kc.Name)
+	// Confirm telemetry config is valid (e.g. Prometheus is installed if it is requested.)
+	promInstalled, err := telemetry.IsPromInstalled(remoteClient, logger)
+	if err != nil {
+		return result.Error(err)
+	}
+	validConfig, err := telemetry.SpecIsValid(mergedSpec, promInstalled)
+	if err != nil {
+		return result.Error(errors.New("could not determine if telemetry config is valid"))
+	}
+	if !validConfig {
+		return result.Error(errors.New("telemetry spec was invalid for this cluster - is Prometheus installed if you have requested it"))
+	}
+	// If Prometheus not installed bail here.
+	if !promInstalled {
+		return result.Continue()
+	}
+	// Determine if we want a cleanup or a resource update.
+	switch {
+	case mergedSpec == nil:
```

Comment:
I do not believe that these can be combined. You propose to combine:
```
case mergedSpec == nil: ...
case mergedSpec.Prometheus == nil: ...
default:
```

Which means that the condition `case mergedSpec.Prometheus.Enabled:` is hit first. In the case that either `mergedSpec` or `mergedSpec.Prometheus` is nil, this will lead to a nil pointer deference.

You need to go step by step to ensure that each of the members exist, that's why the switch statement is there.

Reply:
I don't have a problem implementing it this way if you prefer. It does not affect the behaviour of the reconciliation in either event.